src/core/transaction.c

```diff
@@ -648,11 +648,9 @@ static int transaction_apply(
                 assert(!j->transaction_prev);
                 assert(!j->transaction_next);
 
-                r = hashmap_ensure_allocated(&m->jobs, NULL);
-                if (r < 0)
+                r = hashmap_ensure_put(&m->jobs, NULL, UINT32_TO_PTR(j->id), j);
+                if (r == -ENOMEM)
                         return r;
```

Comment:
Please drop these two lines, and always goto `rollback`.

Reply:
umm makes sense done